ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -15,17 +15,175 @@
+   * this way the results will be associated with some meaningful data.
+   *
+   * e.g. given : [task1_order0 >>= action1 , task2_order1, task3_order0 >>= action2]
+   * we first launch task1 and task3, when both task1 and task3 has finished we execute
+   * action1 and action2 recursively to completion.
+   * Only then we launch task2 and waits for its completion.
+  *)
+end = struct
+  (** structured like a Freer Monad *)
+  type +'a t =
+    | Completed of 'a
+        (** [Completed result] an action that has finished, sometimes refered to as 'pure' *)
+    | Task of API.ref_task * (task_result -> 'a t)
+        (** [Task(task, next)] a task, and a function to call when the task finishes (continuation).
+     * This is an impure action. *)
+
+  (* Constructors *)
+
+  let return x = Completed x
+
+  let ok x = Completed (Result.Ok x)
+
+  let fail e = Completed (Result.Error e)
+
+  let wrap f = try f () with e -> Backtrace.is_important e ; fail e
+
+  let task f = wrap (fun () -> Task (f (), return))
+
+  (* Operations *)
+
+  let rec fmap f = function
+    | Completed x ->
+        (* just map the result through [f] if we already have it *)
+        Completed (f x)
+    | Task (task, task_next) ->
+        (* If everything was immediately available then the result would be:
+       * [f (task_next (result_of_task task))]
+       * However we first need to wait for the task to finish,
+       * and when we receive the result from the task, we need to map it through
+       * [task_next] and then through [f].
+       * The type system helps to ensure that the pipeline below is correct *)
+        Task (task, fun x -> x |> task_next |> fmap f)
+
+
+  module Infix = struct
+    let rec ( >>= ) m k =
+      match m with
+      | Completed x -> k x
+      | Task (task, next) ->
+          (* similar reasoning as above, when we get the result we need to chain the computations,
+         * refer to http://okmij.org/ftp/Computation/free-monad.html for a deeper theoretical explanation *)
+          Task (task, fun x -> next x >>= k)
+  end
+
+  let eval ~__context = function
+    | Completed _ as t -> t
+    | Task (task, next) as t ->
+      match result_of_task ~__context task with
+      | None -> t
+      | Some result ->
+          log_and_ignore_exn (fun () -> Xapi_task.destroy ~__context ~self:task) ;
+          next result
+
+
+  let classify completed lst =
+    List.fold_left
+      (fun (completed, pending, tasks) -> function
+        | k, Completed r -> ((k, r) :: completed, pending, tasks)
+        | (_, Task (task, _)) as t -> (completed, t :: pending, task :: tasks)
+        )
+      (completed, [], []) lst
+
+
+  let parallel ~__context ~rpc ~session_id lst =
+    let rec loop completed lst =
+      (* find any pending tasks we can wait for right now *)
+      match classify completed lst with
+      | completed, [], [] ->
+        debug "All %d parallel actions have completed" (List.length completed);
```

Comment:
The debug messages should indicate the module where they origin. Maybe use `(%s) .. __LOC__` or similar.

Reply:
They'll already have 'xapi_ha_vm_failover' due to the declaration of the Debug module at the top of the file. It's true that it it won't have the line number but very few of our log lines currently have that and this is sufficiently unique that a grep will undoubtedly find this line.